Apache-Beam-Jobs.md

```diff
@@ -227,6 +229,257 @@ error_pcoll = (
+
+  * Did the job run without any problems? How and why do I know?
+  * How much work did the job manage to do?
+  * If the job encountered a problem, what caused it?
+
+Our job is trying to report the total number of states across all explorations, so we need to create `JobRunResult` that holds that information. For this, we can simply use the `as_stdout` helper method:
+
+```python
+def run(self):
+    exp_model_pcoll = (
+        self.pipeline
+        | 'Get all ExplorationModels' >> ndb_io.GetModels(
+            exp_models.ExplorationModel.get_all(deleted=False))
+    )
+
+    state_count_pcoll = (
+        exp_model_pcoll
+        | 'Count states' >> beam.Map(self.get_number_of_states)
+    )
+
+    state_count_sum_pcoll = (
+        state_count_pcoll
+        | 'Sum values' >> beam.CombineGlobally(sum)
+    )
+
+    return state_count_sum_pcoll | beam.Map(job_run_result.JobRunResult.as_stdout)
+```
+
+The method maps every element in a `PCollection` into a `JobRunResult` with the stringified-value as its `stdout`.
+
+With this, our job is fully written! Move on to the next section for instructions on _testing_ your new job.
+
+## Testing Apache Beam Jobs
+
+We'll assume that we've written a fully operational `CountExplorationStatesJob`, and go over the steps in ensuring the job works as intended.
+
+### 1. Inherit from `JobTestBase` and override the class constant `JOB_CLASS`
+
+The current convention is to name your test cases `<JobName>Tests`, but you can create better names if you want to break tests up into groups of specialized focus. For our example, we'll keep things simple.
+
+```python
+class CountExplorationStatesJobTests(test_jobs.JobTestBase):
+
+    JOB_CLASS = CountExplorationStatesJob
+```
+
+### 2. Run assertions using a `assert_job_output_is_*` method
+
+```python
+def test_empty_datastore(self):
+    # Don't add any explorations to the datastore.
+    self.assert_job_output_is_empty()
+
+def test_single_exploration(self):
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e1', 'o1', ['A', 'B', 'C'], ['TextInput'])
+
+    self.assert_job_output_is([
+        job_run_result.JobRunResult(stdout='3'),
+    ])
+
+def test_many_explorations(self):
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e1', 'o1', ['A', 'B', 'C'], ['TextInput'])
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e1', 'o1', ['D', 'E', 'F', 'G', 'H'], ['TextInput'])
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e1', 'o1', ['I', 'J'], ['TextInput'])
+    self.save_new_linear_exp_with_state_names_and_interactions(
+        'e1', 'o1', ['K', 'L', 'M', 'N'], ['TextInput'])
+
+    self.assert_job_output_is([
+        job_run_result.JobRunResult(stdout='14'),
+    ])
+```
+
+Note that `self.assert_job_output_is(...)` and `self.assert_job_output_is_empty()` do as advertised, it runs the job to completion and verifies the result.
```

Comment:
... as advertised -- they run the job to completion and verify the result.

Reply:
Done.